### PR TITLE
Use parameter_types instead of thread_local for test-setup

### PR DESCRIPTION
**Pull request by tifecool**
PR for: #12020 

WIP

Fixes: https://github.com/paritytech/substrate/issues/12020
might close: https://github.com/paritytech/substrate/issues/10479

polkadot address: 1s51dYXziQwTnSnModErKfra3TeEsVipqzY846CLCTEXtHJ

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1746777

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1746778

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1746999

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1746998

File: frame/assets/src/mock.rs [hunk 1]
```diff
@@ -101,44 +101,53 @@ impl Config for Test {
...
-pub(crate) enum Hook {
+pub enum Hook {
 	Died(u32, u64),
 }
-thread_local! {
-	static FROZEN: RefCell<HashMap<(u32, u64), u64>> = RefCell::new(Default::default());
-	static HOOKS: RefCell<Vec<Hook>> = RefCell::new(Default::default());
+parameter_types! {
+	static FROZEN: HashMap<(u32, u64), u64> = Default::default();
+	static HOOKS: Vec<Hook> = Default::default();
```
Comment (kianenigma) [inline frame/assets/src/mock.rs:112]:
```suggestion
	static Frozen: HashMap<(u32, u64), u64> = Default::default();
	static Hooks: Vec<Hook> = Default::default();
```

The macro automatically converts them to SCREAM_CASE under the hood.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Good direction, but: 

1. naming needs fixing
2. Please add https://github.com/paritytech/substrate/pull/12002/files#diff-31d6476b566f793c17f32f9fdefeb487988af58c61ba752d0275b91f6c8aaccaR433 to your PR and use it instead of `:get` and `:set` when appropriate

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1748220

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1748335

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1748405

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1750233

**Comment by tifecool** [comment]
So I'm noticing some of the checks are failing because I haven't changed the variables to camel case, but on changing the variables to camel case I'm getting errors. How do I work around this? @kianenigma

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1750594

File: frame/aura/src/mock.rs [hunk 2]
```diff
@@ -86,26 +85,25 @@ impl pallet_timestamp::Config for Test {
 	type WeightInfo = ();
 }
 
-thread_local! {
-	static DISABLED_VALIDATORS: RefCell<Vec<AuthorityIndex>> = RefCell::new(Default::default());
+parameter_types! {
+	static DISABLED_VALIDATORS: Vec<AuthorityIndex> = Default::default();
```
Comment (tifecool) [inline frame/aura/src/mock.rs:89]: Couldn't convert this to CamelCase because another variable uses the name. What could this be renamed to? @kianenigma

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1754151

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1754150

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1754152

File: frame/aura/src/mock.rs (hunk 2 above)
Comment (kianenigma) [reply frame/aura/src/mock.rs:89]: Just use some other name, like `DisabledValidatorTestValue`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1756360

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1756436

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1756438

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1756477

File: frame/assets/src/mock.rs [hunk 3]
```diff
@@ -101,44 +101,51 @@ impl Config for Test {
...
 pub(crate) fn hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.borrow().clone())
+	Hooks::get().clone()
 }
 
 pub(crate) fn take_hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.take())
+	let result = Hooks::get();
+	Hooks::set(Default::default());
+	result
```
Comment (shawntabrizi) [inline frame/assets/src/mock.rs:148]: no take api? probably we could add one

File: frame/bounties/src/tests.rs [hunk 4]
```diff
@@ -102,8 +101,8 @@ impl pallet_balances::Config for Test {
 	type AccountStore = System;
 	type WeightInfo = ();
 }
-thread_local! {
-	static TEN_TO_FOURTEEN: RefCell<Vec<u128>> = RefCell::new(vec![10,11,12,13,14]);
+parameter_types! {
+	static TentoFourteen: Vec<u128> = vec![10,11,12,13,14];
```
Comment (shawntabrizi) [inline frame/bounties/src/tests.rs:105]:
```suggestion
	static TenToFourteen: Vec<u128> = vec![10,11,12,13,14];
```

Comment (shawntabrizi) [reply frame/bounties/src/tests.rs:105]: is this even used?

File: frame/child-bounties/src/tests.rs [hunk 5]
```diff
@@ -105,8 +104,8 @@ impl pallet_balances::Config for Test {
 	type AccountStore = System;
 	type WeightInfo = ();
 }
-thread_local! {
-	static TEN_TO_FOURTEEN: RefCell<Vec<u128>> = RefCell::new(vec![10,11,12,13,14]);
+parameter_types! {
+	static TenToFourteen: Vec<u128> = vec![10,11,12,13,14];
```
Comment (shawntabrizi) [inline frame/child-bounties/src/tests.rs:108]: same here, is it used?

**Review by shawntabrizi — APPROVED** [review]
looks good to me. thanks!

Please add your polkadot address to the PR if you would like a tip: https://github.com/paritytech/substrate-tip-bot#usage

**Comment by tifecool** [comment]
Hey @shawntabrizi. The PR isn't finished actually. Just wanted all the checks to pass before I slept off yesterday. Will be done probably today or tomorrow.

File: frame/assets/src/mock.rs (hunk 3 above)
Comment (tifecool) [reply frame/assets/src/mock.rs:148]: Ok, might tackle this.

Comment (kianenigma) [reply frame/assets/src/mock.rs:148]: Should be very trivial to make, similar to how I've asked you to add `::mutate`.

File: frame/contracts/src/tests.rs [hunk 6]
```diff
@@ -108,10 +108,11 @@ pub mod test_utils {
 	}
 }
 
-thread_local! {
-	static TEST_EXTENSION: RefCell<TestExtension> = Default::default();
+parameter_types! {
+	static TestExtensionsTestValue: TestExtension = Default::default();
```
Comment (kianenigma) [inline frame/contracts/src/tests.rs:112]: Why is this renamed?

File: frame/contracts/src/tests.rs [hunk 7]
```diff
@@ -339,19 +339,19 @@ impl Convert<Weight, BalanceOf<Self>> for Test {
...
 	pub fn set_filter(filter: fn(&Call) -> bool) {
-		CALL_FILTER.with(|fltr| *fltr.borrow_mut() = filter);
+		CallFilterTestValue::mutate(|fltr| *fltr = filter);
 	}
 }
 
 impl Contains<Call> for TestFilter {
 	fn contains(call: &Call) -> bool {
-		CALL_FILTER.with(|fltr| fltr.borrow()(call))
+		CallFilterTestValue::get()(call)
```
Comment (kianenigma) [inline frame/contracts/src/tests.rs:354]: I don't see why you postfixed everything with `TestValue`? I suggested it _only_ when there is a naming conflict 😅

Comment (tifecool) [reply frame/contracts/src/tests.rs:354]: Wasn't everything, they had conflicts

File: frame/assets/src/mock.rs (hunk 3 above)
Comment (tifecool) [reply frame/assets/src/mock.rs:148]: I think I can implement it directly actually. Made a minor adjustment to mutate so it returns the object after it's been mutated (similar to the `.with` method)

Comment (tifecool) [reply frame/assets/src/mock.rs:148]: Doesn't work, but the method would have to be implemented on Vec, not in the macro, I believe.

**Comment by tifecool** [comment]
I think the PR is done after this.

**Comment by tifecool** [comment]
The PR should be done after this.

**Comment by tifecool** [comment]
Please review @shawntabrizi @kianenigma

File: frame/assets/src/mock.rs [hunk 8]
```diff
@@ -101,44 +101,51 @@ impl Config for Test {
...
 }
 
 pub(crate) fn hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.borrow().clone())
+	Hooks::get().clone()
 }
 
 pub(crate) fn take_hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.take())
+	let result = Hooks::get();
```
Comment (kianenigma) [inline frame/assets/src/mock.rs:146]: You could build a `fn take() -> T` in `parameter_types` as well 🙈

File: frame/contracts/src/tests.rs (hunk 6 above)
Comment (kianenigma) [reply frame/contracts/src/tests.rs:112]: Still a valid question?

File: frame/im-online/src/tests.rs [hunk 9]
```diff
@@ -96,7 +97,8 @@ fn should_report_offline_validators() {
 		advance_session();
 
 		// then
-		let offences = OFFENCES.with(|l| l.replace(vec![]));
+		let offences = OFFENCES::get();
```
Comment (kianenigma) [inline frame/im-online/src/tests.rs:100]: yeah, a `take` function is useful.

File: frame/merkle-mountain-range/src/mock.rs [hunk 10]
```diff
@@ -91,14 +94,14 @@ impl LeafData {
 	}
 }
 
-thread_local! {
-	pub static LEAF_DATA: RefCell<LeafData> = RefCell::new(Default::default());
+parameter_types! {
+	pub static LeafDataTestValue: LeafData = Default::default();
```
Comment (kianenigma) [inline frame/merkle-mountain-range/src/mock.rs:98]: Again, don't see why you renamed this.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Still some unresolved comments.

File: frame/support/src/lib.rs [hunk 11]
```diff
@@ -438,6 +438,14 @@ macro_rules! parameter_types_impl_thread_local {
...
 						[<$name:snake:upper>].with(|v| *v.borrow_mut() = t);
 					}
+
+					/// Mutate the internal value in place.
+					pub fn mutate<R, F: FnOnce(&mut $type) -> R>(mutate: F) -> R{
+						let mut current = Self::get();
+						let result = mutate(&mut current);
+						Self::set(current);
+						result
+					}
```
Comment (kianenigma) [inline frame/support/src/lib.rs:448]:
```suggestion

					/// Mutate the internal value in place.
					pub fn mutate<R, F: FnOnce(&mut $type) -> R>(mutate: F) -> R {
						let mut current = Self::get();
						let result = mutate(&mut current);
						Self::set(current);
						result
					}
					
					/// Get and remove the value. 
					pub fn take() -> $type {
						let current = Self::take();
						Self::put(Default::default());
						current
					}
```

there you go.

File: frame/im-online/src/tests.rs (hunk 9 above)
Comment (tifecool) [reply frame/im-online/src/tests.rs:100]: My issue with implementing a `take` function on the macro is whether or not all types implement `Default::Defaut()`

Comment (tifecool) [reply frame/im-online/src/tests.rs:100]: Seen your implementation. Will implement. Thought there would be issues.

File: frame/support/src/lib.rs (hunk 11 above)
Comment (tifecool) [reply frame/support/src/lib.rs:448]: Noted! Thanks.

File: frame/merkle-mountain-range/src/mock.rs (hunk 10 above)
Comment (tifecool) [reply frame/merkle-mountain-range/src/mock.rs:98]: I can't name it `LeafData` because there's already a struct named `LeafData`

File: frame/contracts/src/tests.rs (hunk 6 above)
Comment (tifecool) [reply frame/contracts/src/tests.rs:112]: There's a struct named `TestExtension`, would cause a conflict.

File: frame/assets/src/mock.rs (hunk 8 above)
Comment (tifecool) [reply frame/assets/src/mock.rs:146]: Done 🫡

Comment (tifecool) [reply frame/assets/src/mock.rs]: Having an issue:
```
the trait bound `for<'r> fn(&'r tests::Call) -> bool: std::default::Default` is not satisfied
   --> frame/contracts/src/tests.rs:342:1
```
How would I go about implementing Default here? For others, I can derive.

**Comment by tifecool** [comment]
Having an issue:
```
the trait bound `for<'r> fn(&'r tests::Call) -> bool: std::default::Default` is not satisfied
   --> frame/contracts/src/tests.rs:342:1
```

File: frame/election-provider-multi-phase/src/mock.rs [hunk 12]
```diff
@@ -241,8 +241,9 @@ impl pallet_balances::Config for Runtime {
 	type WeightInfo = ();
 }
 
-#[derive(Eq, PartialEq, Debug, Clone, Copy)]
```
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/mock.rs:244]: Why is this needed?

File: frame/im-online/src/tests.rs (hunk 9 above)
Comment (kianenigma) [inline frame/im-online/src/tests.rs:100]: Not applied yet?

File: frame/im-online/src/tests.rs [hunk 13]
```diff
@@ -68,15 +68,16 @@ fn should_report_offline_validators() {
...
 		advance_session();
 
 		// when
 		// we end current session and start the next one
 		advance_session();
 
 		// then
-		let offences = OFFENCES.with(|l| l.replace(vec![]));
+		let offences = OFFENCES::get();
+		OFFENCES::mutate(|l| l.clear());
```
Comment (kianenigma) [inline frame/im-online/src/tests.rs:80]:
```suggestion
		let offences = Offences::take();
```

File: frame/offences/src/mock.rs [hunk 14]
```diff
@@ -40,13 +40,12 @@ use sp_staking::{
...
 };
-use std::cell::RefCell;
 
 pub struct OnOffenceHandler;
 
-thread_local! {
-	pub static ON_OFFENCE_PERBILL: RefCell<Vec<Perbill>> = RefCell::new(Default::default());
-	pub static OFFENCE_WEIGHT: RefCell<Weight> = RefCell::new(Default::default());
+parameter_types! {
+	pub static OnOffencePerBill: Vec<Perbill> = Default::default();
```
Comment (kianenigma) [inline frame/offences/src/mock.rs:47]:
```suggestion
	pub static OnOffencePerbill: Vec<Perbill> = Default::default();
```

File: frame/offences/src/mock.rs [hunk 15]
```diff
@@ -58,16 +57,16 @@ impl<Reporter, Offender> offence::OnOffenceHandler<Reporter, Offender, Weight>
 		_offence_session: SessionIndex,
 		_disable_strategy: DisableStrategy,
 	) -> Weight {
-		ON_OFFENCE_PERBILL.with(|f| {
-			*f.borrow_mut() = slash_fraction.to_vec();
+		OnOffencePerBill::mutate(|f| {
```
Comment (kianenigma) [inline frame/offences/src/mock.rs:60]:
```suggestion
		OnOffencePerbill::mutate(|f| {
```

**Review by kianenigma — APPROVED** [review]
LGTM modulo some last nits, please make sure all open comments are addressed.

**Comment by tifecool** [comment]
Wow. Really appreciate the tip! Will round up immediately ❤️

File: frame/election-provider-multi-phase/src/mock.rs (hunk 12 above)
Comment (tifecool) [reply frame/election-provider-multi-phase/src/mock.rs:244]: `fn take` from parameter_types needs a default at compile time.

File: frame/im-online/src/tests.rs (hunk 9 above)
Comment (tifecool) [reply frame/im-online/src/tests.rs:100]: On it.

File: frame/support/src/lib.rs [hunk 16]
```diff
@@ -444,6 +444,21 @@ macro_rules! parameter_types_impl_thread_local {
...
+
+					/// Get and remove the value.
+					pub fn take() -> $type {
+						let current = Self::get();
+						Self::set(Default::default());
```
Comment (bkchr) [inline frame/support/src/lib.rs:459]: Hmm isn't this weird?

Maybe setting it to `$value`, but still weird :see_no_evil:

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
@tifecool please merge master

File: frame/support/src/lib.rs (hunk 16 above)
Comment (tifecool) [reply frame/support/src/lib.rs:459]: I'm not so used to macros, why is it weird? And how can I improve it?

Comment (bkchr) [reply frame/support/src/lib.rs:461]: By weird I mean that it is reset to `Default`.  You specify the "default" value and if you take this, I would expect that you get this value again.

Comment (tifecool) [reply frame/support/src/lib.rs:461]: The default value isn't being specified here though.

Comment (bkchr) [reply frame/support/src/lib.rs:461]: Line 428 you got `$value`. That is the default I'm speaking about.

Comment (tifecool) [reply frame/support/src/lib.rs:461]: Oh, the value got at 458 is different from the default.

**Comment by tifecool** [comment]
@athei please review

**Comment by tifecool** [comment]
>Let's do @bkchr's suggestion in https://github.com/paritytech/substrate/pull/12036#discussion_r962948064 and use $value, not Default.

Ok, let me try.

File: frame/support/src/lib.rs (hunk 16 above)
Comment (tifecool) [reply frame/support/src/lib.rs:461]: I think I understand now. Sorry about the confusion, not too literate with macro syntax.

File: frame/contracts/src/storage/meter.rs [hunk 17]
```diff
@@ -441,32 +441,32 @@ mod tests {
...
 		origin: AccountIdOf<Test>,
 		contract: AccountIdOf<Test>,
 		amount: DepositOf<Test>,
 		terminated: bool,
 	}
 
-	#[derive(Default, Debug, PartialEq, Eq)]
-	struct TestExt {
+	#[derive(Default, Debug, PartialEq, Eq, Clone)]
+	pub struct TestExt {
```
Comment (athei) [inline frame/contracts/src/storage/meter.rs:469]: Why does it need to be public?

**Comment by kianenigma** [comment]
Are we sure this closes https://github.com/paritytech/substrate/issues/10479 too?

File: frame/support/src/lib.rs [hunk 18]
```diff
@@ -446,6 +446,21 @@ macro_rules! parameter_types_impl_thread_local {
...
+
+					/// Mutate the internal value in place.
+					pub fn mutate<R, F: FnOnce(&mut $type) -> R>(mutate: F) -> R{
+						let mut current = Self::get();
+						let result = mutate(&mut current);
+						Self::set(current);
+						result
+					}
+
+					/// Get current value and replace with original set value ($value).
```
Comment (kianenigma) [inline frame/support/src/lib.rs:458]:
```suggestion
					/// Get current value and replace with initial value of the parameter type.
```

**Comment by kianenigma** [comment]
bot merge

**Comment by athei** [comment]
> Are we sure this closes #10479 too?

It still uses thread local under the hood. So I don't think so.

File: frame/contracts/src/exec.rs [hunk 19]
```diff
@@ -1420,8 +1420,8 @@ mod tests {
 		refcount: u64,
 	}
 
-	#[derive(Default)]
-	struct MockLoader {
+	#[derive(Default, Clone)]
+	pub struct MockLoader {
```
[the remark this answers is not in the document]
Comment (tifecool) [reply frame/contracts/src/exec.rs:1424]: parameter_types requires it
